docs/guides/trinity/writing_plugins.rst

```diff
@@ -0,0 +1,188 @@
+
+  We should only choose this type of plugin category if what we are trying to build cannot be built
+  with a :class:`~trinity.extensibility.plugin.BaseIsolatedPlugin`.
+
+We build this kind of plugin subclassing from
+:class:`~trinity.extensibility.plugin.BaseAsyncStopPlugin`. A detailed example can be found in the
+"Building Shared Process Plugins" Guide.
+
+
+Defining plugins
+~~~~~~~~~~~~~~~~
+
+We define a plugin by deriving from either
+:class:`~trinity.extensibility.plugin.BaseMainProcessPlugin`,
+:class:`~trinity.extensibility.plugin.BaseIsolatedPlugin` or 
+:class:`~trinity.extensibility.plugin.BaseAsyncStopPlugin` depending on the kind of plugin that we
+intend to write. For now, we'll stick to :class:`~trinity.extensibility.plugin.BaseIsolatedPlugin`
+which is the most commonly used plugin category.
+
+Every plugin needs to overwrite ``name`` so voilà, here's our first plugin!
+
+.. literalinclude:: ../../../trinity/plugins/examples/peer_count_reporter/plugin.py
+   :language: python
+   :start-after: --START CLASS--
+   :end-before: def configure_parser
+
+Of course that doesn't do anything useful yet, bear with us.
+
+Configuring Command Line Arguments
+----------------------------------
+
+More often than not we want to have control over if or when a plugin should start. Adding
+command-line arguments that are specific to such a plugin, which we then check, validate, and act
+on, is a good way to deal with that. Implementing
+:meth:`~trinity.extensibility.plugin.BasePlugin.configure_parser` enables us to do exactly that.
+
+This method is called when Trinity starts and bootstraps the plugin system, in other words,
+**before** the start of any plugin. It is passed a :class:`~argparse.ArgumentParser` as well as a
+:class:`~argparse._SubParsersAction` which allows it to amend the configuration of Trinity's
+command line arguments in many different ways.
+
+For example, here are adding a boolean flag ``--report-peer-count`` to Trinity.
+
+.. literalinclude:: ../../../trinity/plugins/examples/peer_count_reporter/plugin.py
+   :language: python
+   :pyobject: PeerCountReporterPlugin.configure_parser
+
+To be clear, this does not yet cause our plugin to automatically start if ``--report-peer-count``
+is passed, it simply accepts the parser to be aware of such flag and hence allows us to check for
+its existence later.
+
+.. note::
+
+  For a more advanced example, that also configures a subcommand, checkout the ``trinity attach``
+  plugin.
+
+
+Defining a plugins starting point
+---------------------------------
+
+Every plugin needs to have a well defined starting point. The mechanics of this are slightly
+different for a :class:`~trinity.extensibility.plugin.BaseMainProcessPlugin` but fairly similar
+for the other types of plugins which we'll be focussing on for now.
+
+Plugins need to implement the :meth:`~trinity.extensibility.plugin.BasePlugin._start` method
```

Comment:
`do_start` seems fine.  You could also *invert* these and make `_start` the one that is called internally and `start` the one that the user implements.

Reply:
>You could also invert these

That would work for the implementation as of today where Plugins call `start` themselves but I would still consider both APIs public. It is more clear for the `stop` API that is called from the `PluginManager` hence it is definitely public.

But if you are ok with the `do` prefix then I'll send a PR for that. Thinking about it, it was strange to have a method that is intended to be overwritten to be *private* anyway. That wouldn't even work in most languages (e.g. Java, C# etc)

Reply:
Linking https://github.com/ethereum/py-evm/pull/1440

Reply:
I'm not too concerned with getting this exactly right this first time around since this is an experimental API in alpha software.

Reply:
I do agree in general, but overwriting `_start` / `_stop` (soon to be `do_start` / `do_stop`) are **the** most important APIs when it comes to writing plugins and not being able to properly document / link to them (without hacks!) strikes me as wrong.

Reply:
The long-run ideal solution might be switching from inheritance to composition, but I don't immediately see the quick fix to get there. So, `do_start` is fine as an iteration :+1:

Reply:
@cburgdorf as stated in the other PR, I'm fine with the `do_` prefix on the plugin API but I'm :-1: on immediately extending it to the rest of the codebase.

Reply:
>but I'm  -1 on immediately extending it to the rest of the codebase.

@pipermerriam is this because we need to find a better prefix than `do`. Because, from a language theory it strikes me as plain wrong that we have something like `BaseService._run` as *private* **and** *abstract* at the same time. In fact, I don't know any *statically typed* (which Python isn't but there more we use mypy the more we try to be haha) that would allow this.

In something like Java or C#, if you want subclasses to be able to overwrite a method, or even force them to, then that method can not be private.

And it has real consequences, it means that if we document the `BaseService` like this:

```
.. autoclass:: p2p.service.BaseService
  :members:
``` 

Then, `_run` and `_cleanup`, probably the most important methods of this class, those that need to be overwritten, simply fall out of the documentation because they are incorrectly seen as private. Something that would be impossible in most if not all statically typed language that has the concepts of `private` members and inheritance.

Reply:
I think what we're struggling with here is that a bad architectural design is triggering a negative side-effect of some weird OOP practices (this private/protected method that must be implemented by subclasses).

Two types of needs call for two types of solutions:
1. Urgent feature: patch the side effect by adding `do_` (I think Christoph feels urgency on the plugin docs, so they are in this category)
2. Easy maintenance: attack the root architectural problem, probably with converting inheritance to compositional approach. (I think `BaseService` and other instances in the code base 
are in this category)

Reply:
Added my answer in the PR. https://github.com/ethereum/py-evm/pull/1440#issuecomment-433685007